dipy/viz/actor.py

```diff
@@ -0,0 +1,648 @@
+        one can input grid coordinates and display the slice in space (or grid)
+        coordinates as calculated by the affine parameter.
+
+    """
+    if data.ndim != 3:
+        if data.ndim == 4:
+            if data.shape[3] != 3:
+                raise ValueError('Only RGB 3D arrays are currently supported.')
+            else:
+                nb_components = 3
+        else:
+            raise ValueError('Only 3D arrays are currently supported.')
+    else:
+        nb_components = 1
+
+    if value_range is None:
+        vol = np.interp(data, xp=[data.min(), data.max()], fp=[0, 255])
+    else:
+        vol = np.interp(data, xp=[value_range[0], value_range[1]], fp=[0, 255])
+    vol = vol.astype('uint8')
+
+    im = vtk.vtkImageData()
+    if major_version <= 5:
+        im.SetScalarTypeToUnsignedChar()
+    I, J, K = vol.shape[:3]
+    im.SetDimensions(I, J, K)
+    voxsz = (1., 1., 1.)
+    # im.SetOrigin(0,0,0)
+    im.SetSpacing(voxsz[2], voxsz[0], voxsz[1])
+    if major_version <= 5:
+        im.AllocateScalars()
+        im.SetNumberOfScalarComponents(nb_components)
+    else:
+        im.AllocateScalars(vtk.VTK_UNSIGNED_CHAR, nb_components)
+
+    # copy data
+    # what I do below is the same as what is commented here but much faster
+    # for index in ndindex(vol.shape):
+    #     i, j, k = index
+    #     im.SetScalarComponentFromFloat(i, j, k, 0, vol[i, j, k])
+    vol = np.swapaxes(vol, 0, 2)
+    vol = np.ascontiguousarray(vol)
+
+    if nb_components == 1:
+        vol = vol.ravel()
+    else:
+        vol = np.reshape(vol, [np.prod(vol.shape[:3]), vol.shape[3]])
+
+    uchar_array = numpy_support.numpy_to_vtk(vol, deep=0)
+    im.GetPointData().SetScalars(uchar_array)
+
+    if affine is None:
+        affine = np.eye(4)
+
+    # Set the transform (identity if none given)
+    transform = vtk.vtkTransform()
+    transform_matrix = vtk.vtkMatrix4x4()
+    transform_matrix.DeepCopy((
+        affine[0][0], affine[0][1], affine[0][2], affine[0][3],
+        affine[1][0], affine[1][1], affine[1][2], affine[1][3],
+        affine[2][0], affine[2][1], affine[2][2], affine[2][3],
+        affine[3][0], affine[3][1], affine[3][2], affine[3][3]))
+    transform.SetMatrix(transform_matrix)
+    transform.Inverse()
+
+    # Set the reslicing
+    image_resliced = vtk.vtkImageReslice()
+    set_input(image_resliced, im)
+    image_resliced.SetResliceTransform(transform)
+    image_resliced.AutoCropOutputOn()
+    image_resliced.SetInterpolationModeToLinear()
+    image_resliced.Update()
+
+    if nb_components == 1:
+        if lookup_colormap is None:
+            # Create a black/white lookup table.
+            lut = colormap_lookup_table((0, 255), (0, 0), (0, 0), (0, 1))
+        else:
+            lut = lookup_colormap
+
+    x1, x2, y1, y2, z1, z2 = im.GetExtent()
+    ex1, ex2, ey1, ey2, ez1, ez2 = image_resliced.GetOutput().GetExtent()
+
+    class ImageActor(vtk.vtkImageActor):
```

Comment:
Okay this is a good question. The reasons why often the classes are inside the functions are because a) vtk classes do not allow for multiple inheritance and that makes putting the classes out tricky design-wise, b) it makes testing more complicated. Because the visual tests need to be skipped for most of the bots, c) and more importantly I don't have clear in my mind yet an object oriented design for the actors and widgets that could be ready for prime time. From my experience it is better to be conservative in the design in the beginning and be careful on putting too many objects out too quickly. As we will play more with the new viz module it will be clearer (or not) which objects can go out of the functions and which not. However, from experience right now we have to be more careful than usual. Don't forget that vtk is a very messy library in many ways. So, hiding a lot of its complications inside the functions looks like a good strategy for now.
Nonetheless, in the next PR you will see some grid and container objects which can clearly and safely be written as objects.